src/clj_irods/core.clj

```diff
@@ -77,29 +78,38 @@
   passed in as `sym`. Recommended choice is to call it irods."
   [cfg sym & body]
-  `(let [id# (name (gensym "with-irods-"))
-         jargon-cfg# (or (:jargon-cfg ~cfg) (jargon-cfg (:jargon ~cfg)))
-         jargon-opts# (or (:jargon-opts ~cfg) {})
-         use-jargon# (boolean jargon-cfg#)
-         use-icat# (have-icat)
-         jargon-pool# (or (:combined-pool ~cfg) (:jargon-pool ~cfg) (make-threadpool (str id# "-jargon") (or (:jargon-pool-size ~cfg) 1)))
-         icat-pool#   (or (:combined-pool ~cfg) (:icat-pool ~cfg) (make-threadpool (str id# "-icat") (or (:icat-pool-size ~cfg) 5)))]
-     (try+
-       (maybe-icat-transaction use-icat#
-         (maybe-jargon use-jargon# jargon-cfg# jargon-opts# jargon#
-           (let [~sym {:jargon      jargon#
-                       :jargon-pool jargon-pool#
-                       :icat-pool   icat-pool#
-                       :has-jargon  use-jargon#
-                       :has-icat    use-icat#
-                       :cache       (atom {})}]
-             (do ~@body))))
-       (finally
-         (when-not (:retain-jargon-pool ~cfg) (.shutdown jargon-pool#))
-         (when-not (:retain-icat-pool ~cfg) (.shutdown icat-pool#))))))
+  `(otel/with-span [s# ["with-irods" {:kind :internal}]]
+     (let [id# (name (gensym "with-irods-"))
+           jargon-cfg# (or (:jargon-cfg ~cfg) (jargon-cfg (:jargon ~cfg)))
+           jargon-opts# (or (:jargon-opts ~cfg) {})
+           use-jargon# (boolean jargon-cfg#)
+           use-icat# (have-icat)
+           use-icat-transaction# (and use-icat# (get ~cfg :use-icat-transaction true))
+           jargon-pool# (or (:combined-pool ~cfg) (:jargon-pool ~cfg) (make-threadpool (str id# "-jargon") (or (:jargon-pool-size ~cfg) 1)))
+           icat-pool#   (or (:combined-pool ~cfg) (:icat-pool ~cfg) (make-threadpool (str id# "-icat") (or (:icat-pool-size ~cfg) 5)))]
+       (try+
+         (maybe-icat-transaction use-icat-transaction#
+           (maybe-jargon use-jargon# jargon-cfg# jargon-opts# jargon#
+             (let [~sym {:jargon      jargon#
+                         :jargon-pool jargon-pool#
+                         :icat-pool   icat-pool#
+                         :has-jargon  use-jargon#
+                         :has-icat    use-icat#
+                         :cache       (atom {})}]
+               (otel/with-span [s# ["with-irods body" {:kind :internal}]]
+                 (do ~@body)))))
+         (finally
+           (when-not (:retain-jargon-pool ~cfg) (.shutdown jargon-pool#))
+           (when-not (:retain-icat-pool ~cfg) (.shutdown icat-pool#)))))))
 
 ;; Framework functions to be used to prefer cached values and choose between available data sources.
 
+(defmacro instrumented-delay
+  [& body]
+  `(let [ctxspan# (otel/current-span)]
+     (delay (with-open [_# (otel/span-scope ctxspan#)]
+              (otel/with-span [s2# ["delay body" {:kind :internal}]] ~@body)))))
```

Comment:
Like with the agents, this makes sure the opentelemetry spans end up the right place in the tree.